### ISSUE TITLE
problem with cpp templates

### ISSUE THREAD

**Issue by vegansk** [nim-lang/Nim#4063]
The code: 
```nim
type
  Map {.importcpp: "std::map", header: "<map>".} [T,U] = object

proc initMap[T,U](): Map[T,U] {.importcpp: "std::map<'*0,'*1>(@)", nodecl.}

var x = initMap[int, cstring]()
```

doesn't compile: 
```
Error: internal error: invalid apostrophe type parameter index
```

**Comment by vegansk** [comment]
I can get the first type via `'*0` but it fails on `'*1`

**Comment by Araq** [comment]
By the time we are in the C++ codegen the information about 'T' and 'U' is lost. Type substitutions refer to the **parameters** (which is () ), not the generic types (which is [T,U]).

You need a some workaround.

**Comment by vegansk** [comment]
@Araq, the information about the first generic type is not lost. This is working example:
```nim
type
  Vector {.importcpp: "std::vector", header: "<vector>".} [T] = object
  String {.importcpp: "std::string", header: "<string>".} = object

proc initVector[T](): Vector[T] {.importcpp: "std::vector<'*0>(@)", nodecl.}

var x = initVector[String]()
```

and it produces variable declaration:
```c++
std::vector<std::string>  x_94023;
```

**Comment by vegansk** [comment]
@Araq, the documentation say, that ``0`` is the result type, and ``*`` is the base type of the type. ``initMap`` returns ``Map[T,U]``, so ``'0`` is ``Map[T,U]``, am I right? So, why ``'*0`` is ``T`` in this example? Test:
```nim
type
  Map {.importcpp: "std::map", header: "<map>".} [T,U] = object

proc initMap[T](): Map[T,T] {.importcpp: "std::map<'*0,'*0>(@)", nodecl.}

var x = initMap[int]()
```

generates var x as:
```
std::map<NI, NI>  x_94024;
```

**Comment by Araq** [comment]
Well we need something like ``*0.1``

**Comment by vegansk** [comment]
Interesting solution is shown here: #4093. Fixing that issue also fixes this.

**Comment by vegansk** [comment]
Solved by PR #4094
